### ISSUE TITLE
Go 1.15: broken PBSE2 KDF

### ISSUE THREAD

**Issue by eclipseo** [dvsekhvalnov/jose2go#26]
Since Go 1.15 beta 1, the test suite fails with these errors:

```
Testing    in: /builddir/build/BUILD/jose2go-1.3/_build/src
         PATH: /builddir/build/BUILD/jose2go-1.3/_build/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/sbin
       GOPATH: /builddir/build/BUILD/jose2go-1.3/_build:/usr/share/gocode
  GO111MODULE: off
      command: go test -buildmode pie -compiler gc -ldflags " -X github.com/dvsekhvalnov/jose2go/version=1.3 -extldflags '-Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld '"
      testing: github.com/dvsekhvalnov/jose2go
github.com/dvsekhvalnov/jose2go
invalid sign 'alg' header err= jwt.Decode(): required 'alg' header is missing or of invalid type
missing sign 'alg' header err= jwt.Decode(): required 'alg' header is missing or of invalid type
two phased err= Test error
invalid encrypt 'alg' header err= jwt.Decode(): required 'alg' header is missing or of invalid type
invalid encrypt 'enc' header err= jwt.Decode(): required 'enc' header is missing or of invalid type
missing encrypt 'alg' header err= jwt.Decode(): required 'alg' header is missing or of invalid type
missing encrypt 'enc' header err= jwt.Decode(): required 'enc' header is missing or of invalid type
----------------------------------------------------------------------
FAIL: jose_test.go:1588: TestSuite.TestDecrypt_PBSE2_HS256_A128KW_A128CBC_HS256
jose_test.go:1596:
    //then
    c.Assert(err, IsNil)
... value *errors.errorString = &errors.errorString{s:"aes.KeyUnwrap(): integrity check failed."} ("aes.KeyUnwrap(): integrity check failed.")
----------------------------------------------------------------------
FAIL: jose_test.go:1600: TestSuite.TestDecrypt_PBSE2_HS384_A192KW_A192CBC_HS384
jose_test.go:1608:
    //then
    c.Assert(err, IsNil)
... value *errors.errorString = &errors.errorString{s:"aes.KeyUnwrap(): integrity check failed."} ("aes.KeyUnwrap(): integrity check failed.")
----------------------------------------------------------------------
FAIL: jose_test.go:1612: TestSuite.TestDecrypt_PBSE2_HS512_A256KW_A256CBC_HS512
jose_test.go:1620:
    //then
    c.Assert(err, IsNil)
... value *errors.errorString = &errors.errorString{s:"aes.KeyUnwrap(): integrity check failed."} ("aes.KeyUnwrap(): integrity check failed.")
```

Any chance that you take a look at what causing these issues?

**Comment by dvsekhvalnov** [comment]
Hi @eclipseo , 

thanks for letting me know.  I'll take a look. Or if you have PR - it is always welcome too :)

**Comment by dvsekhvalnov** [comment]
Seems to be PBSE2 KDF is badly broken on Go v1.15 probably due to some changes in 'crypto/hash'. Investingating.

**Comment by dvsekhvalnov** [comment]
Appears to be an issue with new crypto/hmac implementation with respect to `.Reset`: https://github.com/golang/go/blob/master/src/crypto/hmac/hmac.go#L59

will dig more

**Comment by dvsekhvalnov** [comment]
Fixed in 1.15.

@eclipseo when you have a chance, give it a try? All fixes have been merged to `master`.

**Comment by eclipseo** [comment]
Hey thanks, works perfectly in Fedora 33 and Rawhide:

- https://koji.fedoraproject.org/koji/taskinfo?taskID=51690331
- https://koji.fedoraproject.org/koji/taskinfo?taskID=51690210
